README.md
Please see [live/README.txt](live/README.txt).<br>

BTW, "live" is just the deliverable working ("portable") server dir,
the only one you actually need from this package.

If you *don't* yet have a working PHP installation on the PATH:

- Linux: just do install the "php5-cgi" package, that should be the
  least painful way.

- Windows XP or above: just (download and) keep the "php" dir, too.
  (If you move it from the default location, change `set PHPDIR=...`
  in `httphp.cmd` accordingly. (Again, see [live/README.txt](live/README.txt)
  for more details.)


live/CHANGES.txt
TODO:
=====
[REMEMBER TO UPDATE 'VERSION' IN THE SOURCE, TOO!]

! Per-webroot config!

! "Dump config" control request (as the Apache server-config)

! Add cmdline parameters: -c configfile.merge-over-defaults, and 
  -C configfile.replace-defaults

! FIX [header-resend]
PHP Strict Standards:  Only variables should be assigned by reference in xxx.php on line xxx
[1399128991241] Shutting down gracefully. Thanks, bye!
http.js:691
    throw new Error('Can\'t set headers after they are sent.');
          ^
Error: Can't set headers after they are sent.
    at ServerResponse.OutgoingMessage.setHeader (http.js:691:11)
    at Socket.<anonymous> (/mnt/local/WORKAREA/prj/webdev/server/httphp/live/lib/php-cgi-sz.js:197:11)
    at Socket.EventEmitter.emit (events.js:95:17)
    at Socket.<anonymous> (_stream_readable.js:746:14)
    at Socket.EventEmitter.emit (events.js:92:17)
    at emitReadable_ (_stream_readable.js:408:10)
    at emitReadable (_stream_readable.js:404:5)
    at readableAddChunk (_stream_readable.js:165:9)
    at Socket.Readable.push (_stream_readable.js:127:10)
    at Pipe.onread (net.js:528:21)
root@lenovo:/mnt/local/SZABIREPO/tmp/fotok/Pecsvarad/ilyen volt - 2010-apr [from neuro online]# 

! Use the HTTP status phrase set by CGI, instead of ignoring it and using
  the Node.js defaults.
  
! Return the error pages as text/html, rather than plain text.

! Add option to ignore the pre-packaged node.exe for those who have their own.
! Fix: "owered-By: PHP/5.5.11-2tent-type: text/htmlript> location="README.txt" </script>"
  on Linux. Just running in this live dir, and opening http://localhost:88.
  -> ONLY OCCURED ONCE, after freshly apt-updating php-cgi & nodejs, and the
    first php-cgi invokation from httphp since the last Debian reboot.

! [add-alias-dir] -a for "add alias dir" (with regexp support, later)
    Also add the corresponding "ctrl:" command (generalize out the "control commander",
    like apachectl!)
    ! General form: '[-a[ ["]dirname[=urlrootpath["]]]]'
    ! Multiple -a options supported.
    ! '-a' alone means '-a .=`dirname(realpath(.))`'
! File-type handlers (move PHP to php-handler, add js etc.).
! Check cookies & PHP sessions.
! Better 'CTRL:' request handling: add some security token?
  (Since this is meant to be local-only, no need to complicate things too far!)
! FIX backslashes in SCRIPT_NAME, PHP_SELF etc. 
  Seems wrong only for files directly in SERVER_DOC_ROOT (but not in subdirs).
  See e.g. the prj browser at P:\)
! Add a nice replaceable autoindex PHP module! (Design a general PHP addon module
  interface for that! (Preferably just some drop-in dir.))

DONE:
=====

+ NOTE: manual change tracking is being obsoleted; check the git commit 
  history instead!

+ Some testing of HTTP Basic Auth; passing credentials to PHP-CGI 
  (PHP_AUTH_USER, PHP_AUTH_PW).

2014-05-27 - 1.15
+ Change cfg. option INDEXES to INDEX. (Basically to make ON_404_TRY_INDEX 
  more intuitive, at the expense of INDEX becoming less so a bit - but that'd 
  need less explanation, too.)
+ Fix broken HTTP header processing from the CGI stream.
+ Fix broken HTTP response codes set by the CGI program.

2014-04-30
+ Fix httphp.sh to find itself also via symlinked paths.

2014-04-18 - 1.14
+ Log the original request URI before talking about file lookup and mappings!

2014-04-13 - 1.13
+ Moved cfg. to separate file (default.cfg).
+ Moved SERVER_XXX props into SERVER_CFG.XXX.
+ Added SERVER_CFG.ON_404_TRY_INDEX = true/false/"filename"

2014-04-12 - 1.12
+ Replace PHP5ts.dll with a vanilla PHP5.dll (currently 5.4.26).
+ Improved the docs.

2014-04-12 - 1.11
+ Log graceful shutdown (exit(), SIGINT, SIGTERM etc.).

2014-04-12 - 1.10
+ Some files dos2unixed.
+ Fix: Silent exit on bind EACCES (e.g. trying to bind to a port under 1000
  on Linux). Also add catch-all error reporting for unknown cases.

2014-04-9 - 1.09
+ "No PHP support" is logged too early, before the "Server started" line. 
  (In fact the server started message is too late!)
  -> Uhh-ohh, that's because missing PHP support is detected in the startup 
   shell script (on Windows), not by the server itself!...
   * Changed "HTTPHP" prefix to "HTTPHP.cmd".

2014-04-02 - 1.08
+ Added logic to HTTPHP.cmd for locating a PHP-CGI .exe to use (on Windows).
+ Added index.php to the live dir (index.html redirects to that index.php, 
  which redirects to README.txt, basically as a "secret" PHP test at startup.)

2014-04-02 - 1.07
+ Sigh, finally gave in and "officially" integrated the minimal Windows PHP 
  example crap lurking around the rest of the package from the beginning.
  (Just to make it a self-containing survival kit for people not having a PHP 
  installation yet. The dir layout is still pretty lame, it's just hanging 
  in the air, making it slightly problematic for some people to "install", 
  I guess. Well, I've got this gift of making enemies, what can I do?)

2014-03-31 - 1.06
+ [php-multislash-internals]
  PHP internal ?=PHPxxx URIs used to fail with surplus slashes (like //index.php).
  See also [url.parse-crash].
  + Check if this is the same for other servers, too!
  -> No, LAMP is fine; Apache normalizes the slashes in PHP_SELF & SCRIPT_NAME
    (which may be the key!), so now I also do the same.

+ [url.parse-crash]
  Crash when double slashes in the URL: localhost// (or localhost//index.html).
  (Curiously, it does *NOT* crash for localhost/// or localhost/.)
  httphp.js:194
  r reqpath = decodeURIComponent(Url.parse(reqpath, false, true).path.replace(/\
                                                                    ^
  TypeError: Cannot call method 'replace' of null
    at Server.<anonymous> (P:\webdev\server\httphp\live\httphp.js:194:72)
    at Server.EventEmitter.emit (events.js:98:17)
    at HTTPParser.parser.onIncoming (http.js:2108:12)

test/web/index.php
This should not be seen unless explicitly requesting index.php!
<hr>
<?php
require("php/index.php");
//phpinfo();
?>


test/web/php/form.php
<pre>
<?php
echo '$_REQUEST == ';
print_r($_REQUEST);
?>
</pre>
<form method="POST">
<input type="text" name="in" value="<?= $_REQUEST['submit']=='ok' ? $_REQUEST['in'] : ''?>">
<input type="submit" name="submit" value="ok">
<input type="submit" name="submit" value="clear">
</form>


test/web/php/index.php
<?php echo 'Hi from PHP!<br>'; ?>
<pre>
<div style="border: 1px solid gray;">
<?php
require 'request.php';
require 'server_vars.php';
?>
</div>
</pre>
<hr>
<a href="form.php">A (PHP-backed) FORM...</a>
<hr>
<?php
phpinfo(); 
?>


live/lib/php-cgi-sz.js
/**
Based on php-cgi 0.2.0
See also: CGI RFC: http://tools.ietf.org/html

Changes by Sz.:

! See the Git commit log for all the recent changes. 
  Some old ones:

+ Disabled bogus detectBinary().
+ Added error (ENOENT) handling for spawn().
+ Disabled bogus PATH_INFO and PATH_TRANSLATED.
+ Fixed PHP startup: added missing REDIRECT_STATUS.
+ Fixed query string handling to support special PHP "?=PHPxxxx" params.
  (Disabled QS-parsing and tentatively removed unnecessary(?) urlencoding.)
+ Fixed response body for non-text data.
+ Confusing 'paramsForRequest' renamed to 'prepareCGIEnv'.
+ 'newRequestParams' renamed to 'getRequestVars'.
+ Changed (simplified & extended) the serveResponse() API.
+ Renamed 'serveResponse' to 'process'.
+ Added missing require("path").
*/
var OS = require('os')
var Path = require('path')
var Url = require('url')
var spawn = require('child_process').spawn


//http://social.msdn.microsoft.com/Forums/vstudio/en-US/15514c1a-b6a1-44f5-a06c-9b029c4164d7/searching-a-byte-array-for-a-pattern-of-bytes?forum=csharpgeneral
function find(array, pattern)
{
    if (!array)   return -1
    if (!pattern) return -1

    var found = -1
    for (var i = 0; i <= array.length - pattern.length; i++)
    {
        found = i
        for (var j = 0; j < pattern.length; j++)
        {
            if (array[i + j] != pattern[j])
            {
                found = -1
                break
            }
        }
        if (found != -1) break
    }
    return found
}


var Exports = {
	bin: "php-cgi",

	env: {
		 'SERVER_SOFTWARE':"nodejs"
		,'SERVER_PROTOCOL':"HTTP/1.1"
		,'GATEWAY_INTERFACE':"CGI/1.1"
		,'SERVER_NAME':OS.hostname()
		,'REDIRECT_STATUS_ENV':0
	},
	setEnv: function(env) {
		for(var e in env) {
			Exports.env[e] = env[e];
		}
	},
	getRequestVars: function() {
		var reqEnv = {};
		for(var keys = Object.keys(Exports.env), l = keys.length; l; --l)	{
		   reqEnv[ keys[l-1] ] = Exports.env[ keys[l-1] ];
		}
		return reqEnv;
	},
	prepareCGIEnv: function(scriptfile, req, reqEnv) {
		if (typeof(reqEnv) == "undefined") reqEnv = Exports.getRequestVars();
		
//!!sz: WHY does 'false' work below?!
//!!	requri = Url.parse(req.url,true);
		requri = Url.parse(req.url,false);
		//set environment variables for this request
		reqEnv['SCRIPT_NAME']     = scriptfile;
//sz: This would not work, as PHP assumes it to be a relative path:
//		reqEnv['SCRIPT_NAME'] = scriptfile;
		reqEnv['PATH_INFO']       = ""; //! UNSUPPORTED
		reqEnv['PATH_TRANSLATED'] = ""; //! UNSUPPORTED

//sz: CGI paranoia kludge required by PHP-CGI
//http://stackoverflow.com/questions/7047426/call-php-from-virtual-custom-web-server
//http://woozle.org/~neale/papers/php-cgi.html
		reqEnv['REDIRECT_STATUS'] = 1;
//sz: This is used as a fallback, if SCRIPT_NAME (i.e. URI.pathname) is empty (e.g. http://host/)
		reqEnv['SCRIPT_FILENAME'] = Path.resolve(Path.join(reqEnv['DOCUMENT_ROOT'], scriptfile));
//!!console.log("SCRIPT_FILENAME: " + reqEnv['SCRIPT_FILENAME'])

//sz: console.log("DOCUMENT_ROOT: " + reqEnv['DOCUMENT_ROOT'])
//sz: console.log("requri.pathname: " + requri.pathname)

//sz:
		reqEnv['QUERY_STRING'] = requri.query || '';
//console.log("QS = "+reqEnv['QUERY_STRING']);
//!!		reqEnv['QUERY_STRING'] = '';
//!!		for(var p in requri.query) {
//!!			reqEnv['QUERY_STRING'] += p+"="+encodeURIComponent(requri.query[p])+"&";
//!!		}
		reqEnv['REQUEST_METHOD'] = req.method;
		
		//add request headers, "User-Agent" -> "HTTP_USER_AGENT"
		for (var header in req.headers) {
			reqEnv['HTTP_' + header.toUpperCase().split("-").join("_")] = req.headers[header];
		}
		//copy in additional special headers..
		if ('content-length' in req.headers) {
			reqEnv['CONTENT_LENGTH'] = req.headers['content-length'];
		}
		if ('content-type' in req.headers) {
			reqEnv['CONTENT_TYPE'] = req.headers['content-type'];
		}
		if ('authorization' in req.headers) {
			reqEnv['AUTH_TYPE'] = req.headers.authorization.split(' ')[0];
		}
/* This would be needed later for "mod-php" mode, if basic auth support gets built in:

		// Turn "Authorization: Basic QWxhZGRpbjpvcGVuIHNlc2FtZQ==" to PHP_AUTH_USER, PHP_AUTH_PW
		// (http://en.wikipedia.org/wiki/Basic_access_authentication, http://www.php.net/manual/en/features.http-auth.php)
		if (req.headers.authorization) {
//! Already added:	reqEnv['HTTP_AUTHORIZATION'] = req.headers.authorization;
			var parts = req.headers.authorization.split(" ");
			if (parts[0] != "Basic") {
//!!				log_warning("Unsupported authentication method: '"+parts[0]+"'")
			} else {
				reqEnv['AUTH_TYPE'] = parts[0];
				parts = new Buffer(parts[1], 'base64').toString('ascii');
				parts = parts.split(":");
				reqEnv['PHP_AUTH_USER'] = parts[0];
				reqEnv['PHP_AUTH_PW']   = parts[1];
				console.log('PHP_AUTH_USER = "' + parts[0] + '"');
				console.log('PHP_AUTH_PW = "' + parts[1] + '"');
			}
		}
*/
		return reqEnv;
	},

	/** sz: WHAT DOES THIS COMMENT MEAN?
	* "This is an automatic function, will add a function you can override later on."
	*/
	process: function(scriptfile, req, res, end_callback, params) {
		if (typeof(end_callback) == "object") { params = end_callback; end_callback = null; }
		if (typeof(params) == "undefined") params = {};
		var reqEnv = Exports.prepareCGIEnv(scriptfile, req);
		var cgi = spawn(Exports.bin, [], {
			  env: reqEnv,
			  stdio: 'pipe'
		});
		var http_status = { code: 200, text: "OK" }; // might be overridden from PHP
		var error_msg  = "CGI error"

		// spawn() would pass errors here:
		cgi.on('error', function(e) {
			switch (e.code) {
			case 'ENOENT':
				error_msg = "Couldn't find CGI executable '"+Exports.bin+"'."
				break
			case 'OK':	// Huh?! (See [cgi-ok-error].)
				return
			default:
				error_msg = "'" + e.code + "' while trying to launch CGI executable '"+Exports.bin+"'."
				break
			}
			console.log("CGI: "+ error_msg)
			res.writeHead(500, "Internal server error: " + error_msg)

			if (typeof(end_callback) == "function")
				end_callback(500, error_msg) // this must close the response stream!
			else
				res.end()

		})


		req.pipe(cgi.stdin);
		if (params['sterr']) {
			cgi.stderr.on('data',params.sterr);
		} else {
			cgi.stderr.on('data',function(data) {
				console.log("CGI error: "+ data.toString());
			});
		}	
		var headersSent = false;
		cgi.stdout.on('data', function(data) {		
			if (headersSent) {
				//stream data to browser as soon as it is available.
				//console.log(data.toString());
				res.write(data);
			} else {
				
//!!sz: Added support for non-text data
//http://docs.nodejitsu.com/articles/advanced/buffers/how-to-use-buffers
				var headers_end = find(data, new Buffer("\r\n\r\n"));
				if (headers_end != -1) {
//console.log("*** CGI headers found, ending at: " + headers_end);
//console.log(data.toString("utf-8", 0, headers_end));
					var lines = data.toString("utf-8", 0, headers_end).split("\r\n");
					// Set the headers...
					for (var l=0; l<lines.length; l++) {
						var hdrname, hdrvalue, hdrname_end = lines[l].indexOf(":");
						if ((hdrname_end > 0) 
							&& (hdrname = lines[l].substring(0, hdrname_end).trim()) )
						{
							hdrvalue = lines[l].substr(hdrname_end+1).trim() //! hdrname_end+1
							res.setHeader(hdrname, hdrvalue);
							// Is the status set by the CGI app?
							if (hdrname == "Status") {
								http_status.code = parseInt(hdrvalue);
								http_status.text = null; //!! just force default phrase for now
							}
						}
					}

					if (!res.getHeader("content-length")) {
					//stream the output
						res.setHeader('Transfer-Encoding', 'chunked');
					}
					res.writeHead(http_status.code, http_status.text);
					headersSent = true;
				}
//				if ((type = res.getHeader("content-type")) && 
//					type.substring(0, 4) == "text") {
//					res.write(lines.slice(l+1).join('\r\n'));
//				} else {
					res.write(data.slice(headers_end + 4));
//				}
			}
		});
		cgi.stdout.on('end',function() { 
			if (typeof(end_callback) == "function") {
				end_callback(http_status.code, http_status.text) //! the callback must close the response stream
			} else {
				res.end()
			}
		});
	}
}

module.exports = Exports

live/httphp.js
// Product:
var NAME = "HTTPHP"
var VERSION = "1.15+"

var SERVER_CFG = require('./default.cfg')

// URI -> dir ("alias") map. Will be set after processing the args!
//!! (How this relates to the legacy SERVER_CFG.DOC_ROOT is not 100% clear yet. 
//!! For now '/' -> SERVER_CFG.DOC_ROOT is the only default alias map entry, and
//!! SERVER_CFG.DOC_ROOT is, unfortunately, also used directly at several places
//!! "the" (only) legacy docroot value.)
var dirmap = {
	'/': SERVER_CFG.DOC_ROOT,
}

// Stealing from: http://trac.nginx.org/nginx/browser/nginx/conf/mime.types
var map_ext_to_content_type = {
	".html":        "text/html",
	".htm":		"text/html",
	".js":		"application/javascript",
	".css":		"text/css",
	".jpg":		"image/jpeg",
	".jpeg":	"image/jpeg",
	".png":		"image/png",
	".gif":		"image/gif",
	".ico":		"image/x-icon",
	".txt":		"text/plain",
}

// Will set this after params processing:
var SERVER_URL_PREFIX = 'UNINITIALIZED!'

//---------------------------------------------------------------------------
// Helpers, utils. etc...
//---------------------------------------------------------------------------

function show_help() {
	console.log(NAME + "v"+VERSION+" - Instant low-tech PHP server for local dirs")
	console.log("")       
	console.log("httphp           : Serve the current dir as http://localhost/")
	console.log("        -h       : Show this help (and exit).")
	console.log("        -p NUM   : Serve on port NUM instead of 80.")
	console.log("        -d DIR   : Serve DIR instead of '.'")
	console.log("        -i FILES : Try the given (comma-separated) file names")
	console.log("                   (in the given order) as default index files")
	console.log("                   (instead of `index.html,index.php`)")
}


function log(msg) {
	var LOGPREFIX = ""	//"HTTPHP: "
	var TIMESTAMP = "["+Date.now()+"] "
	console.log(LOGPREFIX + TIMESTAMP + msg)
}

function log_http(request, urlpath, localpath, result, note) {
	log(request.method +" "+ request.url +" [mapped to: "+ localpath +"] - "+ result + " "+ note)
}

function log_err(msg) {
	log("ERROR: "+ msg)
}
function log_warn(msg) {
	if (SERVER_CFG.LOG_FILTER.indexOf('warn') == -1)
		log("WARNING: "+ msg)
}
function log_notice(msg) {
	if (SERVER_CFG.LOG_FILTER.indexOf('notice') == -1)
		log("notice: "+ msg)
}
function log_debug(msg) {
	if (SERVER_CFG.LOG_FILTER.indexOf('debug') == -1)
		log(">> DEBUG <<: "+ msg)
}


//---------------------------------------------------------------------------
// main...
//---------------------------------------------------------------------------
var Http = require('http')
var Path = require("path")  
var Url = require("url")
var Fs = require("fs")
//var Util = require("util")
var phpCGI = require("./lib/php-cgi-sz");

log("HTTPHP Server v" +VERSION+ " at "+__dirname+" started.")

// Options to override the defaults:
// -p port
// -d webdir
// -i default index files (file1,...)
var param_of = null
process.argv.forEach(function (arg, i, array) {
	switch (arg) {

	case '-h':
		show_help()
	        process.exit()

	case '-p': param_of = '-p'
		return
	case '-d': param_of = '-d'
		return
	case '-i': param_of = '-i'
		return
	}

	switch (param_of) {
	case '-p':
		SERVER_CFG.PORT = arg
		log_notice("SERVER.PORT set to '" + SERVER_CFG.PORT + "'")
		param_of = null
		break
	case '-d':
		SERVER_CFG.DOC_ROOT = arg
		log_notice("SERVER.DOC_ROOT set to '" + SERVER_CFG.DOC_ROOT + "'")
		param_of = null
		break
	case '-i':
		SERVER_CFG.INDEX = arg
		log_notice("SERVER.INDEX set to '" + SERVER_CFG.INDEX + "'")
		param_of = null
		break
	}			
});
if (param_of) {
	err("malformed command-line (near '"+ param_of +"')");
	return
}

// Set this for the request handler
//! NOTE: THESE CAN'T BE RECONFIGURED RUN-TIME, so enough to set it once.
SERVER_URL_PREFIX = SERVER_CFG.PROTOCOL + "://" + SERVER_CFG.HOST + ":" + SERVER_CFG.PORT


function respond_404(request, response, reqpath, file_to_serve_fullpath) {
	if (typeof(file_to_serve_fullpath) == 'undefined') file_to_serve_fullpath = '?'
	response.writeHeader(404, {"Content-Type": "text/plain"})
	response.write("404 Not Found\n")
	response.end()
	log_http(request, reqpath, file_to_serve_fullpath, 404, "Not Found")
}

// Check dirs in the dir map (for each or for a specific url)
function check_alias(dirmap, specific_url) {
//!! Since we'll have to handle filesystem errors for each individual request
//!! anyway, this checking here is nothing more than just a vague example.
//!! Then it's better to leave a broken alias on, since the user may want to
//!! fix it while the server is running.

	var url//, keys = dirmap.keys
	for (url in dirmap) {
		if (typeof(specific_url) === 'undefined' 
			|| url == specific_url) { // Make this check real: trim, substr. etc!
			var dir = dirmap[url]
			log_debug("Checking dir '" + dir + "'")
			if (!Fs.existsSync(dir)) {
				log_err("Cannot serve non-existing dir '"+ dir +"', exiting...")
				return
			}
		}
	}
}

// 0. Precondition: each key in dirmap must end with a trailing slash.
// 1. Take each alias 'a' from dirmap, in their set order.
// 2. If uri_path.indexOf(a) == 0, then return dirmap[a].
// 3. Else return dirmap['/'].
/*
function get_root_dir_for_uri_path(uri_path) {
	if (uri_prefix in dirmap && uri_path.indexOf(a) == 0)
		return dirmap[uri_prefix]
	else
		return dirmap['/']
}
*/

/*
function get_index(request, response, reqpath, file_to_serve_fullpath) {

	log_debug("Falling back to dir. index on missing direct match...")
	return false
}
*/

// Based on: http://www.hongkiat.com/blog/node-js-server-side-javascript/
var server = Http.createServer(function(request, response) {
	//! sz: hostname is stripped from request.url

 	//!!HOW ABOUT DECODING THE QS TOO? Whose repsponsibility is that?
 	//!!Currently the PHP-CGI handler does it alone (from the raw request)

	var canonical_url = (SERVER_URL_PREFIX + request.url)
						.replace(/([^:\/])[\/]+([^\/])/g, '$1/$2') // - this is for:
						// [url.parse-crash] - Path.extname would fail if path starts with '//'!
						// [php-multislash-internals] - multiple slashes in PHP_SELF & SCRIPT_NAME break phpinfo()

	var parsed_uri = Url.parse(
						canonical_url/* still URL-encoded! */,
						false/* strip QS! */, 
						false/* canonical_url is absolute & unambiguous */
					)
	var reqpath = parsed_uri.pathname || '/'
	var file_to_serve = ''
	var file_to_serve_fullpath = ''
	var file_ext = ''

	log_notice("Request for '" + request.url + "' received.")

	log_debug("reqpath before decoding: " + reqpath)
	reqpath = decodeURIComponent(reqpath.replace(/\+/g, ' ')) //https://groups.google.com/forum/#!topic/nodejs/8P7GZqBw0xg
	file_to_serve = reqpath

//	log_debug('request.host: '+request.host)
	log_debug("canonical_url: " + canonical_url)
	log_debug("parsed_uri.host: " + parsed_uri.host)
	log_debug("parsed_uri.hostname: " + parsed_uri.hostname)
	log_debug("parsed_uri.pathname: " + parsed_uri.pathname)
	log_debug("parsed_uri.path: " + parsed_uri.path)
	log_debug("URI-path: " + parsed_uri.pathname)
//	log_debug("URI-path + QS: " + parsed_uri.path) //! QS has been stripped off
	log_debug("reqpath decoded: " + reqpath)
	log_debug("file_to_serve: " + file_to_serve)

	// Handle internal command requests (!!EXPERIMENTAL HACK YET!)
	if (request.url.indexOf("ctrl:stop!") > -1) {
		log_notice("STOP! (Requested by ["+request.url+"] (!!source tracking not implemented!!)");

		response.writeHeader(200)
		response.write("Cheers!", "binary")
		response.end()

		// Give some time to the server for responding to the kill request...
		setTimeout(function () {
			        process.exit()
		}, 1000);

		return
	}

	//
	// Locate the resource (generally a file) to serve...
	//
	file_to_serve_fullpath = SERVER_CFG.DOC_ROOT + file_to_serve
	log_debug('file_to_serve_fullpath: ' + file_to_serve_fullpath)

	var file_state = null
	try {
		file_state = Fs.statSync(file_to_serve_fullpath)//!!, function(err, stats) {
	} catch(err) {
		if (err.code == 'ENOENT') {
			//!! The URL may map to an existing file, but that's not necessarily an error!
			//!! BUT NOW, FOR DEBUGGING:
			if (!SERVER_CFG.ON_404_TRY_INDEX) {
				respond_404(request, response, reqpath, file_to_serve_fullpath)
				return
			}
			
			log_notice("Falling back to dir. index on 404 (as configured)...")
			// Fall through to index handling!
			// Also: hack the index file list, if 'ON_404_TRY_INDEX' is itself a file name:
			if (typeof(SERVER_CFG.ON_404_TRY_INDEX) === 'string') {
				SERVER_CFG.INDEX = SERVER_CFG.ON_404_TRY_INDEX + ',' + SERVER_CFG.INDEX
			}
			
		} else {
			response.writeHeader(500, {"Content-Type": "text/plain"})
			response.write(err + "\n")
			log_http(request, reqpath, file_to_serve_fullpath, 500, "File stat() failed")
			response.end()
			return
		}
	}

	// Check if the request points to an existing dir, and
	// then append an index file if one exists in that dir.
	// If not, bail out with 404.
	if (file_state && file_state.isDirectory() || !file_state && SERVER_CFG.ON_404_TRY_INDEX) { //!NOTE: file_state is null if not found

		var INDIRECT_INDEX = null

		//! Don't chop it if it's a real dir request!
		if (!file_state && SERVER_CFG.ON_404_TRY_INDEX) {
			reqpath = Path.dirname(reqpath)
		}

		SERVER_CFG.INDEX.split(",").some(function(index) {
			file_to_serve          = Path.join(reqpath, index.trim())
			file_to_serve_fullpath = Path.join(SERVER_CFG.DOC_ROOT, file_to_serve)
			if (Fs.existsSync(file_to_serve_fullpath)) {
				INDIRECT_INDEX = file_to_serve_fullpath
				log_debug("using index file: '" + file_to_serve_fullpath +"'")
				return true
			} else {
				log_debug("index file not found: '" + file_to_serve_fullpath +"'")
			}
		})

		if (!INDIRECT_INDEX) {
			log_err("no directory index found")
			respond_404(request, response, reqpath, file_to_serve_fullpath)
			return
		}
	}

	log_debug("file_to_serve: " + file_to_serve)

	// NOTE: file_state.isSymbolicLink() transparently handled by stat()
	if (INDIRECT_INDEX || file_state.isFile()) { //!NOTE: file_state may be null

		// Based on the php-cgi README:
		// Just use 'path' instead of Url.parse(request.url, true)

		file_ext = Path.extname(file_to_serve)
		log_debug("filename suffix: " + file_ext)

		switch (file_ext) {

			case ".php":
				phpCGI.env['REQUEST_URI'] = request.url; // PHP-specific
				phpCGI.env['DOCUMENT_ROOT'] = SERVER_CFG.DOC_ROOT + Path.sep;
				phpCGI.env['SERVER_PORT'] = SERVER_CFG.PORT;
				phpCGI.env['SERVER_NAME'] = request.headers.host;

				phpCGI.process(file_to_serve, request, response, function(statuscode, errmsg) {
					log_http(request, reqpath, file_to_serve_fullpath, statuscode, "(via PHP-CGI)")
					switch (statuscode) {
					case 500:
						response.write("HTTP error 500 ("+errmsg+")")
						break
					}
					response.end()
				});

				break;

			default:
				Fs.readFile(file_to_serve_fullpath, "binary", function(err, file) {

					if (!err) {
						response.setHeader("Content-Type", map_ext_to_content_type[file_ext]);
						response.writeHeader(200)
						response.write(file, "binary")
						response.end()
						log_http(request, reqpath, file_to_serve_fullpath, 200, "OK")

					} else if (err.code == 'ENOENT') {
						log_debug("Should this 404 ever occur here?!") //!!?
						respond_404(request, response, reqpath, file_to_serve_fullpath)

					} else {

						response.writeHeader(500, {"Content-Type": "text/plain"})
						response.write(err + "\n")
						response.end()
						log_http(request, reqpath, file_to_serve_fullpath, 500, "File Error")
					}
				})
		}

        } else {
		log_debug("Should this other 404 ever occur here?!") //!!?
		respond_404(request, response, reqpath, file_to_serve_fullpath)
	}

//!!    })
})


process.on('exit', function() {
  log("Shutting down gracefully. Thanks, bye!");
})

process.on('SIGINT', function() {
  log_notice("Exiting on SIGINT (Ctrl+C)...");
  process.exit();
})

process.on('SIGTERM', function() {
  log_notice("Exiting on SIGTERM (kill -15)...");
  process.exit();
})


var server_listening_ok = false
var kill_request_sent = false

server.on('error',function(e){
	if (e.code == 'EADDRINUSE') {
		log_warn("Port "+ SERVER_CFG.PORT +" is already used. Trying to take over...");

		//http://nodejs.org/api/http.html#http_http_get_options_callback
		Http.get("http://"+SERVER_CFG.HOST+":"+SERVER_CFG.PORT+"/?ctrl:stop!", function(res) {
			log_notice("Response to the STOP req.: " + res.statusCode);
		}).on('error', function(e) {
			// Connection reset is expected if we kill the server ;)
			if (e.code != 'ECONNRESET') {
				log_err("'"+ e.message +"' while trying to reclaim server port.");
			}
		});

		setTimeout(function () {
			// Let's retry, regardless of the response:
			server.listen(SERVER_CFG.PORT, SERVER_CFG.HOST)
			// Wait for it to probably finish:
			setTimeout(function () {
				if (!server_listening_ok) {
					log_err("Couldn't reclaim server port.");
				        process.exit()
				}
			}, 500);
		}, 2000);
	} else if (e.code == 'EACCES') {
		log_err("No permission to bind to port " + SERVER_CFG.PORT + ".");
	} else {
		log_err("Unknown fatal error: " + e.code + ".");
	}
})

server.on('listening',function(){
	server_listening_ok = true
	log("Listening on http://"+ SERVER_CFG.HOST + ":" + SERVER_CFG.PORT 
		+ "/ (serving "+ Path.resolve(SERVER_CFG.DOC_ROOT) +").")

	//Just for myself: what exactly these other ways of printing are/were?
	//Util.puts("Server Running on port " + SERVER_CFG.PORT);
	// Sys.puts("Sys.puts vs. Util.puts vs. console.log?...");  
})

server.listen(SERVER_CFG.PORT, SERVER_CFG.HOST)
